src/Knp/Menu/Matcher/Voter/RouteVoter.php

```diff
@@ -11,27 +12,54 @@
 class RouteVoter implements VoterInterface
 {
     /**
-     * @var Request
+     * @var RequestStack|null
+     */
+    private $requestStack;
+
+    /**
+     * @var Request|null
      */
     private $request;
 
-    public function __construct(Request $request = null)
+    public function __construct($requestStack = null)
     {
-        $this->request = $request;
+        if ($requestStack instanceof RequestStack) {
+            $this->requestStack = $requestStack;
+        } elseif ($requestStack instanceof Request) {
+            // BC layer for the old API of the class
+            $this->request = $requestStack;
+        } elseif (null !== $requestStack) {
+            throw new \InvalidArgumentException('The first argument of %s must be null, a RequestStack or a Request. %s given', __CLASS__, is_object($requestStack) ? get_class($requestStack) :  gettype($requestStack));
+        }
     }
 
-    public function setRequest(Request $request)
+    /**
+     * Sets the request against which the menu should be matched.
+     *
+     * This Request is ignored in case a RequestStack is passed in the constructor. This setter
+     * is meant to provide compatibility with Symfony versions before 2.4 when the RequestStack
+     * was not available.
+     *
+     * @param Request|null $request
+     */
+    public function setRequest(Request $request = null)
     {
         $this->request = $request;
     }
 
     public function matchItem(ItemInterface $item)
     {
-        if (null === $this->request) {
+        if (null !== $this->requestStack) {
+            $request = $this->requestStack->getCurrentRequest();
```

Comment:
i think this should be getMasterRequest(). i don't see a use case for basing this on a sub request.